test/backends.py

```diff
@@ -0,0 +1,217 @@
+    def _gen_jax_wrapper(base_cls):
+        class JaxPyroWrapper(base_cls):
+            def _pyro_make_array(self, res_list):
+                """This works around (e.g.) numpy.exp not working with object
+                arrays of :mod:`numpy` scalars. It defaults to making object arrays,
+                however if an array consists of all scalars, it makes a "plain old"
+                :class:`numpy.ndarray`.
+
+                See ``this numpy bug <https://github.com/numpy/numpy/issues/18004>`__
+                for more context.
+                """
+
+                from numbers import Number
+                # Needed to play nicely with Jax, which frequently creates
+                # arrays of shape () when handed numbers
+                all_numbers = all(
+                    isinstance(e, Number)
+                    or (isinstance(e, self.usr_np.ndarray) and e.shape == ())
+                    for e in res_list)
+
+                if all_numbers:
+                    return self.usr_np.array(res_list, dtype=self.usr_np.float64)
+
+                result = self.usr_np.empty_like(res_list, dtype=object,
+                                                shape=(len(res_list),))
+
+                # 'result[:] = res_list' may look tempting, however:
+                # https://github.com/numpy/numpy/issues/16564
+                for idx in range(len(res_list)):
+                    result[idx] = res_list[idx]
+
+                return result
+
+            def _pyro_norm(self, argument, normord):
+                """This works around numpy.linalg norm not working with scalars.
+
+                If the argument is a regular ole number, it uses :func:`numpy.abs`,
+                otherwise it uses ``usr_np.linalg.norm``.
+                """
+                # Wrap norm for scalars
+                from numbers import Number
+                if isinstance(argument, Number):
+                    return self.usr_np.abs(argument)
+                # Needed to play nicely with Jax, which frequently creates
+                # arrays of shape () when handed numbers
+                if (isinstance(argument, self.usr_np.ndarray)
+                        and argument.shape == ()):
+                    return self.usr_np.abs(argument)
+                return self.usr_np.linalg.norm(argument, normord)
+
+        return JaxPyroWrapper
+
+    @staticmethod
+    def generate_code(sol: ct.Solution, name: str):
+        return pyrometheus.codegen.python.gen_thermochem_code(sol=sol)
+
+    @staticmethod
+    def extract_interface(module, mechname: str, user_np):
+        base_cls = module.Thermochemistry
+
+        if user_np == jnp:
+            base_cls = PythonBackend._gen_jax_wrapper(base_cls)
+
+        return base_cls(user_np)
+
+
+class CppBackend(Backend):
+    def get_name():
+        return "cpp"
+
+    def fix_arg(self, x):
+        if isinstance(x, Iterable) and not isinstance(x, str):
+            return self.pyro.VectorDouble(x)
```

Comment:
Neat. Can we work with other array types?

Reply:
Ok, good to know. I don't think need to do more than what already is in there at this stage. Just wanted to know if you could envision this generalizing to other array types.

Reply:
Also, some documentation here would help.

Reply:
@ecisneros8 The documentation for `fix_arg` is in the abstract base class. Otherwise these two lines are pretty self-explanatory.

Reply:
Sorry that it wasn't clear. I meant to explain, more broadly in the documentation (not particularly here), what a `pyro.VectorDouble` is. We can move forward without it at this point.

Reply:
But it is good to leave a trace of these discussions.